core/include/gnuradio-4.0/Block.hpp

```diff
@@ -1337,290 +1426,119 @@ class Block : public lifecycle::StateMachine<Derived>, //
-        if constexpr (StrideControl::kEnabled) {
-            if (stride != 0UL) {
-                const bool firstTimeStride = stride_counter == 0;
-                if (firstTimeStride) {
-                    // sample processing are done as usual, portsStatus.in_samples samples will be processed
-                    if (stride.value > stride_counter + ports_status.in_available) { // stride can not be consumed at once -> start stride_counter
-                        stride_counter += ports_status.in_available;
-                        nSamplesToConsume = ports_status.in_available;
-                    } else { // if the stride can be consumed at once -> no stride_counter is needed
-                        stride_counter    = 0;
-                        nSamplesToConsume = stride.value;
-                    }
-                } else {
-                    // |====================|...|====================|==============----| -> ====== is the stride
-                    //   ^first                    ^we are here (1)  or ^here (2)
-                    // if it is not the "first time" stride -> just consume (1) all samples or (2) missing rest of the samples
-                    // forward tags but no additional sample processing are done ->return
-                    if (stride.value > stride_counter + ports_status.in_available) {
-                        stride_counter += ports_status.in_available;
-                        nSamplesToConsume = ports_status.in_available;
-                    } else { // stride is at the end -> reset stride_counter
-                        nSamplesToConsume = stride.value - stride_counter;
-                        stride_counter    = 0;
-                    }
-                    const auto inputSpans = prepareInputStreams();
-                    const bool success    = consumeReaders(self(), nSamplesToConsume);
-                    return { requested_work, nSamplesToConsume, success ? work::Status::OK : work::Status::ERROR };
-                }
-            }
-        }
-
-        const auto inputSpans   = prepareInputStreams();
-        auto       writersTuple = prepareOutputStreams();
+        // for non-bulk processing, the processed span has to be limited to the first sample if it contains a tag s.t. the tag is not applied to every sample
+        const bool limitByFirstTag = (!HasProcessBulkFunction<Derived> && HasProcessOneFunction<Derived>) &&hasTag;
 
+        // call the block implementation's work function
+        const auto   inputSpans   = prepareStreams(inputPorts<PortType::STREAM>(&self()), limitByFirstTag ? 1 : decimatedIn);
+        auto         writersTuple = prepareStreams(outputPorts<PortType::STREAM>(&self()), limitByFirstTag ? 1 : decimatedOut);
+        work::Status ret;
+        std::size_t  processed    = limitByFirstTag ? 1 : decimatedIn;
+        std::size_t  processedOut = limitByFirstTag ? 1 : decimatedOut;
```

Comment:
```suggestion
        std::size_t  nSamplesToProcessOut = limitByFirstTag ? 1 : decimatedOut;
```

Reply:
These are the only 2 vairables where I didn't manage to make them const, otherwise i'd agree with the rename, but these are also updated with the result of the processing...
What i'm missing here is the ability to define an uninitialized const variable and then have the compiler check that it will be assigned exactly once in every branch... Java does that, but in c++ the only way to do get something like that is to wrap it in a lambda and return from there, but that would make the code a lot less straightforward to read...